### PR TITLE
Fix: libcrmcommon: Escape newlines and tabs in XML attribute values

### PR DESCRIPTION
**Pull request by nrwahl2**
Regression introduced by c6d9cea9, sort of.

Before that commit, newlines and tabs were escaped as `"\n"` and `"\t"`, respectively. These are not valid XML escape sequences.

We made the decision not to escape newlines and tabs when we added the `pcmk__xml_escape()` function, with the caveat that the "right" answer was unclear:

> If not, the XML will be more visually understandable, but conformant XML parsers will treat the newline as a space (for example, if something extracts the text and displays it to the user). If we do, the parser case may be closer to our original intent, but the XML is harder to read, and it may not be appropriate (for example, the parsing tool might prefer to do line wrapping of a textual paragraph on its own)."
https://github.com/ClusterLabs/pacemaker/pull/3323#discussion_r1474790296

However, now if an attribute value contains a newline or tab, it gets replaced by a space character.

In some cases, an attribute value may contain newlines or tabs that need to be preserved. So here, we replace them with their respective character references.

---

When displaying XML attributes in text output format, escape special characters in the attribute value. (An attribute name isn't allowed to have special characters.) This ensures the entire attribute value stays on one line, that a newline doesn't get converted to a space, and that it doesn't have (for example) a double quote inside the attribute value.
    
`attrd_updater` and `crm_attribute` use this output method. It turns out that the `openstack-info` resource agent sets a multiline attribute (which I consider a bug, but it's longstanding behavior). The `openstack-virtual-ip` resource agent then depends on being able to parse it from the `attrd_updater` text output as **one line with no spaces**.
    
If there's a literal newline character in the displayed value, then there's whitespace in the output, breaking the agent's parsing.
    
(The `openstack-info` agent should set the attribute to a saner value, and the `openstack-virtual-ip` agent should parse the XML output instead of the text output, but here we are.)

### CODE REVIEW

File: lib/pacemaker/pcmk_output.c [hunk 1]
```diff
@@ -2141,6 +2141,13 @@ attribute_default(pcmk__output_t *out, va_list args)
 
     GString *s = g_string_sized_new(50);
 
+    char *value_esc = NULL;
+
+    if (pcmk__xml_needs_escape(value, true)) {
+        value_esc = pcmk__xml_escape(value, true);
+        value = value_esc;
+    }
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_output.c:2149]: Set a multi-line attribute:
```
$ sudo attrd_updater -N laptop -n openstack_ports -U "$(echo -e "subnet_id_a\nsubnet_id_b:port_id_1,subnet_id_c\nsubnet_id_d:port_id_2")"
```

Query before:
```
$ sudo attrd_updater -N laptop -n openstack_ports -Q
name="openstack_ports" host="laptop" value="subnet_id_a
subnet_id_b:port_id_1,subnet_id_c
subnet_id_d:port_id_2"
```

Query after:
```
$ sudo attrd_updater -N laptop -n openstack_ports -Q 
name="openstack_ports" host="laptop" value="subnet_id_a&#x0A;subnet_id_b:port_id_1,subnet_id_c&#x0A;subnet_id_d:port_id_2"
```

Note that this change is conservative: it applies only to outputting attributes in text format. Also, dumping XML to text (i.e., to a string representation of the XML) already escaped special characters as appropriate throughout the document.

Any other path through which an attribute value might be represented as text, will not escape newlines (etc.) if it didn't escape them before.

**Comment by clumens** [comment]
This looks good to me.

File: lib/pacemaker/pcmk_output.c (hunk 1 above)
Comment (nrwahl2) [reply lib/pacemaker/pcmk_output.c:2149]: BTW, the values may be fetched from `attrd` rather than from the CIB XML, which is why they're not escaped when we get them. (Another possibility would be to escape them when we store them in attrd, but I'm inclined to keep the characters the way they were received unless escapes are needed for XML or display.)

**Comment by nrwahl2** [comment]
This needs to be updated to work with `crm_attribute()`, which has its own `attribute_text()` output function. That gets values directly from the CIB, so we need to take care not to escape something that's already been escaped -- in other words, don't escape an ampersand if it's the beginning of a character reference or entity reference. Which is a pain.

**Comment by nrwahl2** [comment]
Set a multi-line attribute:
```
$ sudo attrd_updater -N laptop -n openstack_ports -U "$(echo -e "subnet_id_a\nsubnet_id_b:port_id_1,subnet_id_c\nsubnet_id_d:port_id_2")"
```

Query before c6d9cea9 (the newlines are "\n" because that's also how they're "escaped" in the XML; "\n" doesn't have special meaning as an escape sequence in XML):
```
$ sudo attrd_updater -N laptop -n openstack_ports -Q
name="openstack_ports" host="laptop" value="subnet_id_a\nsubnet_id_b:port_id_1,subnet_id_c\nsubnet_id_d:port_id_2"

$ sudo attrd_updater -N laptop -n openstack_ports -Q --output-as=xml
<pacemaker-result api-version="2.35" request="attrd_updater -N laptop -n openstack_ports -Q --output-as=xml">
  <attribute name="openstack_ports" value="subnet_id_a\nsubnet_id_b:port_id_1,subnet_id_c\nsubnet_id_d:port_id_2" host="laptop"/>
  <status code="0" message="OK"/>
</pacemaker-result>
```

Query before "Escape newlines and tabs in XML attribute values" commit:
```
$ sudo attrd_updater -N laptop -n openstack_ports -Q
name="openstack_ports" host="laptop" value="subnet_id_a subnet_id_b:port_id_1,subnet_id_c subnet_id_d:port_id_2"

$ sudo attrd_updater -N laptop -n openstack_ports -Q --output-as=xml
<pacemaker-result api-version="2.35" request="attrd_updater -N laptop -n openstack_ports -Q --output-as=xml">
  <attribute name="openstack_ports" value="subnet_id_a subnet_id_b:port_id_1,subnet_id_c subnet_id_d:port_id_2" host="laptop"/>
  <status code="0" message="OK"/>
</pacemaker-result>
```

Query before the "Escape special chars in XML attribute text output" commit:
```
$ sudo attrd_updater -N laptop -n openstack_ports -Q
name="openstack_ports" host="laptop" value="subnet_id_a
subnet_id_b:port_id_1,subnet_id_c
subnet_id_d:port_id_2"

$ sudo attrd_updater -N laptop -n openstack_ports -Q --output-as=xml
<pacemaker-result api-version="2.35" request="attrd_updater -N laptop -n openstack_ports -Q --output-as=xml">
  <attribute name="openstack_ports" value="subnet_id_a&#10;subnet_id_b:port_id_1,subnet_id_c&#10;subnet_id_d:port_id_2" host="laptop"/>
  <status code="0" message="OK"/>
</pacemaker-result>
```

Query after all these commits (the newlines are character references, which is the *semantically valid* way they're now represented in the XML):
```
$ sudo attrd_updater -N laptop -n openstack_ports -Q 
name="openstack_ports" host="laptop" value="subnet_id_a&#x0A;subnet_id_b:port_id_1,subnet_id_c&#x0A;subnet_id_d:port_id_2"

$ sudo attrd_updater -N laptop -n openstack_ports -Q --output-as=xml
<pacemaker-result api-version="2.35" request="attrd_updater -N laptop -n openstack_ports -Q --output-as=xml">
  <attribute name="openstack_ports" value="subnet_id_a&#10;subnet_id_b:port_id_1,subnet_id_c&#10;subnet_id_d:port_id_2" host="laptop"/>
  <status code="0" message="OK"/>
</pacemaker-result>
```

Note that this change is conservative: it applies only to outputting attributes in text format via the "attribute" message. Also, dumping XML to text (i.e., to a string representation of the XML) already escaped special characters as appropriate throughout the document.

Any other path through which an attribute value might be represented as text, will not escape newlines (etc.) if it didn't escape them before.

If we feel that we need to be especially paranoid about backward compatibility, then we can make `attrd_updater` and `crm_attribute` display newlines as `\n` instead of as `&#x0A;`. However:
* that's going to be ugly to implement (we'll need either a separate escape function or an extra boolean and some special cases in `pcmk__xml_escape/pcmk__xml_needs_escape`)
* `\n` is neither a literal newline nor a valid XML character reference, so it has no special meaning.

**Comment by nrwahl2** [comment]
All commits until "Don't escape when creating XML" are from #3386. I figure that PR is likely to be merged soon, and I don't want to have a rebase conflict in the new `pcmk__xe_set_content()`, so I pulled those commits over here.

**Comment by nrwahl2** [comment]
I added a few refactors on top for `pcmk__xml_escape()` and related functions, that had been on my to-do list. (I would've made some big mistakes if we didn't have unit tests :) )

File: lib/fencing/st_lha.c
Comment (nrwahl2) [inline lib/fencing/st_lha.c:1]: Since this commit looks unrelated: its purpose is to facilitate the next commit, where `pcmk__xml_escape()` returns a `gchar *`.

File: lib/common/xml.c [hunk 2]
```diff
@@ -1071,18 +1071,15 @@ pcmk__xml_needs_escape(const char *text, bool escape_quote)
...
                 return true;
             case '"':
-                if (escape_quote) {
-                    return true;
-                }
-                break;
             case '\n':
             case '\t':
-                // Don't escape newline or tab
+                if (for_attr) {
```
Comment (nrwahl2) [inline lib/common/xml.c:1042]: However, as I think about this more, I think it makes sense for newlines and tabs to be escaped in text nodes too. The main counterpoint was that it makes the XML harder to read. That's true, but as we said back then, conformant XML parsers will convert the newlines and tabs to single spaces, thus changing the output.

XML isn't meant for human eyes except for while we're developing. Any client tool should be passing it to an XML parser. That parser will then convert a `&#x0A;` to a newline, which will show up as a real newline if the parser then displays the text.

Comment (nrwahl2) [reply lib/common/xml.c:1042]: Never mind. An XML processor replaces newlines and tabs with spaces only during attribute-value normalization. In other words, this does NOT happen to text nodes. Literal newlines and tabs are the same as escaped ones.

Comment (nrwahl2) [reply lib/common/xml.c:1042]: Looking at all this again has also made me think 714b4204 was unnecessary -- XML isn't really for human eyes -- but it should be fine, as non-ASCII characters aren't "special" characters for an XML processor.

File: lib/pacemaker/pcmk_output.c [hunk 3]
```diff
@@ -2141,6 +2141,13 @@ attribute_default(pcmk__output_t *out, va_list args)
 
     GString *s = g_string_sized_new(50);
 
+    char *value_esc = NULL;
+
+    if (pcmk__xml_needs_escape(value, true)) {
+        value_esc = pcmk__xml_escape(value, true);
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_output.c:2147]: Fabio is concerned that some agent or tool may depend on the literal `"\n"` (backslash followed by newline, NOT an escape character) that's in the output as of Pacemaker 2.1.7.

Prior to this commit, we output a literal newline (with line break) for text output objects if the attribute value has a newline, so the value gets split over multiple lines. This commit causes a newline to be printed as `"&#x0A;"` by `attrd_updater` and `crm_attribute` for text output objects.

That is my preference, because it's actually a meaningful character and because printing `"\n"` further complicates the implementation. (`pcmk__xml_escape()` never prints `"\n"` for newline anymore.) We could print `"\n"`, however.

Comment (kgaillot) [reply lib/pacemaker/pcmk_output.c:2147]: How were we escaping this to begin with, since this is not used for XML output?

File: tools/crm_attribute.c [hunk 4]
```diff
@@ -63,6 +63,13 @@ attribute_text(pcmk__output_t *out, va_list args)
     const char *value = va_arg(args, const char *);
     const char *host G_GNUC_UNUSED = va_arg(args, const char *);
 
+    char *value_esc = NULL;
+
+    if (pcmk__xml_needs_escape(value, true)) {
+        value_esc = pcmk__xml_escape(value, true);
+        value = value_esc;
+    }
```
Comment (kgaillot) [inline tools/crm_attribute.c:71]: ditto

File: lib/common/xml.c [hunk 5]
```diff
@@ -1104,12 +1103,14 @@ pcmk__xml_needs_escape(const char *text, bool for_attr)
                 }
                 break;
             default:
-                if ((text[index] < 0x20) || (text[index] >= 0x7F)) {
+                if ((*text < 0x20) || (*text >= 0x7F)) {
```
Comment (kgaillot) [inline lib/common/xml.c:1064]: We should be able to use `isprint()` here

File: lib/common/xml.c [hunk 6]
```diff
@@ -1169,62 +1127,73 @@ pcmk__xml_needs_escape(const char *text, bool for_attr)
...
                 if (for_attr) {
-                    copy = replace_text(copy, &index, &length, "&#x09;");
+                    g_string_append(copy, "&#x09;");
+                } else {
+                    g_string_append_c(copy, *text);
                 }
                 break;
             default:
-                if ((copy[index] < 0x20) || (copy[index] >= 0x7F)) {
+                if ((*text < 0x20) || (*text >= 0x7F)) {
```
Comment (kgaillot) [inline lib/common/xml.c:1185]: ditto

File: lib/pacemaker/pcmk_output.c (hunk 3 above)
Comment (nrwahl2) [reply lib/pacemaker/pcmk_output.c:2147]: Good question. I don't see a direct call to `crm_xml_escape()`.

I'm assuming something is sending a query, getting an XML response, parsing it for some reason, getting the attr value, and printing it. That would explain the behavior in both cases (once something becomes "\n", it never goes back to a newline; and "&#x0A;" is parsed back to a newline).

Trying to trace how that might be happening.

Comment (nrwahl2) [reply lib/pacemaker/pcmk_output.c:2147]: `pcmk__ipc_prepare_iov()` on the IPC client side dumps the request XML to a string before sending it. That turns the newline into whatever we use to escape a newline, if anything. Then we parse that string on the IPC server side before setting the attribute.

So a newline set by a CLI tool used to be escaped to `"\n"` before set time. In `main`, it doesn't get escaped, so the dump/parse sequence transforms it to a single space. As of this commit, it gets *properly* escaped, so the dump/parse sequence preserves the newline, and it's ultimately dumped as `"&#x0A;"`

File: lib/common/xml.c (hunk 5 above)
Comment (nrwahl2) [reply lib/common/xml.c:1064]: That's handy

**Comment by nrwahl2** [comment]
new push uses `isprint()`

File: lib/pacemaker/pcmk_output.c (hunk 3 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_output.c:2147]: The XML output looks ideal now. The text output is most correct before "Escape special chars in XML attribute text output" (i.e. the text output is identical to what the user added), though of course OpenStack can't handle that. It doesn't make sense to put XML entities in text output, so I'm leaning to Fabio's recommendation to keep showing backslash-n in that case. We can add a COMPAT comment to eventually switch to the correct text output (hopefully we can get OpenStack to fix it to handle both formats, then switch at 3.1).

Comment (nrwahl2) [reply lib/pacemaker/pcmk_output.c:2147]: > The text output is most correct before "Escape special chars in XML attribute text output" (i.e. the text output is identical to what the user added)

I agree

> backslash-n in that case

Alright, I'll do it as cleanly as I can. It will probably involve an extra strdup() in the "attribute" message formatter, since I don't want to add and maintain another argument to `pcmk__xml_escape()`. Should be fine since it's rarely called.

> hopefully we can get OpenStack to fix it to handle both formats, then switch at 3.1

Yeah it's a trivial fix within the agents. I demonstrated a fix in the agent that works for both formats without having to check feature set or anything. It hasn't been implemented because the push has been to "fix" Pacemaker.

Fabio is concerned that there may be other, currently unknown agents that depend on the `"\n"`. I feel like we've gotta make the switch at some point though.

He is also concerned that the recent changes in escaping (particularly within the XML) may affect rolling upgrades. I told him I don't think they would; once an attribute is set, its value won't change just because Pacemaker was upgraded to a version that escapes differently. Let me know if you feel differently. The only possible edge case that comes to mind is implicit resources, and even that seems unlikely to change based on escaping changes.

Comment (nrwahl2) [reply lib/pacemaker/pcmk_output.c:2147]: > I don't want to add and maintain another argument to `pcmk__xml_escape()`

Still true but I think that might be the easier approach

**Comment by nrwahl2** [comment]
Filed https://github.com/ClusterLabs/resource-agents/pull/1929 against resource-agents

File: lib/common/xml.c [hunk 7]
```diff
@@ -1168,62 +1126,72 @@ pcmk__xml_needs_escape(const char *text, bool for_attr)
  *       \c xmlEncodeEntitiesReentrant() and \c xmlEncodeSpecialChars().
  *       However, their escaping is incomplete. See:
  *       https://discourse.gnome.org/t/intended-use-of-xmlencodeentitiesreentrant-vs-xmlencodespecialchars/19252
+ * \note The caller is responsible for freeing the return value using
+ *       \c g_free().
  */
-char *
+gchar *
```
Comment (kgaillot) [inline lib/common/xml.c:1119]: For some reason `attribute_{default,text}()` flipped back to using `char*`

**Review by kgaillot — COMMENTED** [review]
nice

File: lib/common/xml.c [hunk 8]
```diff
@@ -1046,23 +1046,48 @@ pcmk__xml_needs_escape(const char *text, enum pcmk__xml_escape_type type)
...
+
+            case pcmk__xml_escape_attr_pretty:
+                switch (*text) {
+                    case '\n':
+                    case '\r':
+                    case '\t':
+                    case '"':
+                        return true;
+                    default:
+                        break;
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/xml.c:1090]: I'd rather not add another enum value for the sake of warning if we accidentally use it... I don't remember static analysis caring in that direction, but rather in the direction of wanting an exhaustive switch statement if there's no default case. Nothing complained in CI.

I'd rather `CRM_CHECK()` the type before the loop and return false. That approach would carry over to `pcmk__xml_escape()`: we certainly don't want to warn on every character that we append in the loop, and we can't return `pcmk__str_copy()` after we enter the loop because we might have skipped UTF-8 characters at the beginning.

File: lib/common/xml.c [hunk 9]
```diff
@@ -1111,44 +1136,74 @@ pcmk__xml_escape(const char *text, enum pcmk__xml_escape_type type)
...
-            case '\n':
-                if (type == pcmk__xml_escape_attr) {
-                    g_string_append(copy, "&#x0A;");
-                } else {
-                    g_string_append_c(copy, *text);
+        switch (type) {
+            case pcmk__xml_escape_text:
+                switch (*text) {
+                    case '<':
+                        g_string_append(copy, "&lt;");
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/xml.c:1143]: Yeah I can do that in `crmcommon_private.h` unless you prefer `xml_names_internal.h`. It will only be LT, GT, AMP, and QUOT. We could do special case character references (like `"&#x0A;" for newline), but they would be used only within the unit tests.

File: lib/common/xml.c (hunk 7 above)
Comment (nrwahl2) [reply lib/common/xml.c:1119]: Whoops, thanks! I shuffled commits around to make the "text newline" fix the last commit. Previously, it was before `pcmk__xml_escape()` changed to `gchar *`.

File: lib/common/xml.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/xml.c:1090]: or `CRM_ASSERT()` since it's purely internal and that's definitely a programming error on our side

[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/xml.c:1090]: That would be fine. The issue is that enum variables are just integers and not limited to specified values. So if you want to use a default case to check for unspecified values, static analysis will complain that no values can match the default. Adding an unknown/invalid value and letting default catch it along with unspecified values is a good way around that.

I assume you're thinking of asserting that the value is >= the first value and <= the last. That should work too. I have seen a corner case where the value was unsigned for some reason and testing >= 0 got a complaint that it would always be true.

File: lib/common/xml.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/xml.c]: private header is fine

File: lib/common/xml.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/xml.c:1090]: I did it even more simply (see current commits):
```
default:
    CRM_ASSERT(false);
    break;
```

Neither Coverity nor cppcheck complained about it

**Review by kgaillot — APPROVED** [review]
Looks good to me, let me know if you're ready to merge

Might want to add Fixes T799

**Comment by nrwahl2** [comment]
> Looks good to me, let me know if you're ready to merge
> 
> Might want to add Fixes T799

Done, and I'm ready
